util/generate_dict/generate_dicom_dict.py

```diff
@@ -54,11 +54,81 @@
 from pydicom.values import converters
 
 
-_PKG_DIRECTORY = Path(__file__).parent.parent.parent / "pydicom"
+_PKG_DIRECTORY = Path(__file__).parent.parent.parent / "src" / "pydicom"
 PYDICOM_DICT_FILENAME = _PKG_DIRECTORY / "_dicom_dict.py"
-MAIN_DICT_NAME = "DicomDictionary: Dict[int, Tuple[str, str, str, str, str]]"
-MASK_DICT_NAME = "RepeatersDictionary: Dict[str, Tuple[str, str, str, str, str]]"
-BR = "{https://docbook.org/ns/docbook}"
+PYDICOM_DATASET_STUB = _PKG_DIRECTORY / "stubs" / "dataset.pyi"
+MAIN_DICT_NAME = "DicomDictionary: dict[int, tuple[str, str, str, str, str]]"
+MASK_DICT_NAME = "RepeatersDictionary: dict[str, tuple[str, str, str, str, str]]"
+BR = "{http://docbook.org/ns/docbook}"
+
+
+def write_dataset_stub(attributes):
+    """Update the stub/dataset.pyi file."""
+    autogen_comment = "    # Only auto-generated class attributes below"
+    exclusions = [
+        # File meta elements
+        "FileMetaInformationGroupLength",
+        "FileMetaInformationVersion",
+        "MediaStorageSOPClassUID",
+        "MediaStorageSOPInstanceUID",
+        "TransferSyntaxUID",
+        "ImplementationClassUID",
+        "ImplementationVersionName",
+        "SourceApplicationEntityTitle",
+        "SendingApplicationEntityTitle",
+        "ReceivingApplicationEntityTitle",
+        "SourcePresentationAddress",
+        "ReceivingPresentationAddress",
+        "PrivateInformationCreatorUID",
+        "PrivateInformation",
+        # Image Pixel
```

Comment:
Why not generate these automatically? Do you need them before they are generated?

Reply:
Oh yes, the Type 3 ones can be left for the autogen. The rest are Type 1 so I put them in manually to take advantage of the narrower type.

Actually, hold on. `mypy` will check the types in the `stubs` directory but doesn't actually seem to be using them.